k3k-kubelet/kubelet.go

```diff
@@ -0,0 +1,274 @@
+package main
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"net/http"
+	"os"
+	"time"
+
+	certutil "github.com/rancher/dynamiclistener/cert"
+	"github.com/rancher/k3k/k3k-kubelet/provider"
+	"github.com/rancher/k3k/pkg/apis/k3k.io/v1alpha1"
+	"github.com/rancher/k3k/pkg/controller"
+	"github.com/rancher/k3k/pkg/controller/cluster/server"
+	"github.com/rancher/k3k/pkg/controller/cluster/server/bootstrap"
+	"github.com/rancher/k3k/pkg/controller/kubeconfig"
+	"github.com/virtual-kubelet/virtual-kubelet/log"
+	"github.com/virtual-kubelet/virtual-kubelet/node"
+	"github.com/virtual-kubelet/virtual-kubelet/node/nodeutil"
+	"go.uber.org/zap"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/client-go/kubernetes"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+	"k8s.io/client-go/util/retry"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var Scheme = runtime.NewScheme()
+
+func init() {
+	_ = clientgoscheme.AddToScheme(Scheme)
+	_ = v1alpha1.AddToScheme(Scheme)
+}
+
+type kubelet struct {
+	name       string
+	port       int
+	hostConfig *rest.Config
+	hostClient ctrlruntimeclient.Client
+	virtClient kubernetes.Interface
+	node       *nodeutil.Node
+}
+
+func newKubelet(c *config) (*kubelet, error) {
+	hostConfig, err := clientcmd.BuildConfigFromFlags("", c.HostConfigPath)
+	if err != nil {
+		return nil, err
+	}
+
+	hostClient, err := ctrlruntimeclient.New(hostConfig, ctrlruntimeclient.Options{
+		Scheme: Scheme,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	virtConfig, err := virtRestConfig(context.Background(), c.VirtualConfigPath, hostClient, c.ClusterName, c.ClusterNamespace)
+	if err != nil {
+		return nil, err
+	}
+
+	virtClient, err := kubernetes.NewForConfig(virtConfig)
+	if err != nil {
+		return nil, err
+	}
+	return &kubelet{
+		name:       c.NodeName,
+		hostConfig: hostConfig,
+		hostClient: hostClient,
+		virtClient: virtClient,
+	}, nil
+}
+
+func (k *kubelet) RegisterNode(srvPort, namespace, name, hostname string) error {
+	providerFunc := k.newProviderFunc(namespace, name, hostname)
+	nodeOpts := k.nodeOpts(srvPort, namespace, name, hostname)
+
+	var err error
+	k.node, err = nodeutil.NewNode(k.name, providerFunc, nodeutil.WithClient(k.virtClient), nodeOpts)
+	if err != nil {
+		return fmt.Errorf("unable to start kubelet: %v", err)
+	}
+	return nil
+}
+
+func (k *kubelet) Start(ctx context.Context) {
+	go func() {
+		ctx := context.Background()
+		logger, err := zap.NewProduction()
+		if err != nil {
+			fmt.Printf("unable to create logger: %s", err.Error())
+			os.Exit(1)
+		}
+		wrapped := LogWrapper{
+			*logger.Sugar(),
+		}
+		ctx = log.WithLogger(ctx, &wrapped)
+		err = k.node.Run(ctx)
```

Comment:
```suggestion
		if err := k.node.Run(ctx)
```

These err values should be scope locked.